Plugins/Events/Events/SkillEvents.cpp

```diff
@@ -0,0 +1,63 @@
+#include "Events/SkillEvents.hpp"
+#include "API/CNWSCreature.hpp"
+#include "API/Functions.hpp"
+#include "Events.hpp"
+#include "Utils.hpp"
+
+namespace Events {
+
+using namespace NWNXLib;
+
+static Hooking::FunctionHook* m_UseSkillHook = nullptr;
+
+SkillEvents::SkillEvents(ViewPtr<Services::HooksProxy> hooker)
+{
+    hooker->RequestExclusiveHook<API::Functions::CNWSCreature__UseSkill, int32_t, API::CNWSCreature*, uint8_t, uint8_t, NWNXLib::API::Types::ObjectID, 
+        NWNXLib::API::Vector, NWNXLib::API::Types::ObjectID, NWNXLib::API::Types::ObjectID, int32_t>(&UseSkillHook);
+
+    m_UseSkillHook = hooker->FindHookByAddress(API::Functions::CNWSCreature__UseSkill);
+}
+
+int32_t SkillEvents::UseSkillHook(
+    API::CNWSCreature* thisPtr,  
+    uint8_t skill,
+    uint8_t subSkill,
+    API::Types::ObjectID target,
+    NWNXLib::API::Vector targetPosition,
+    API::Types::ObjectID area,
+    API::Types::ObjectID usedItem,
+    int32_t activePropertyIndex)
+{
+    int32_t retVal;
+
+    Events::PushEventData("SKILL_ID", std::to_string(skill));
```

Comment:
I actually like the lambda approach as opposed to duplicated code. When you push all this stuff, you can wrap it as:
```C
auto Push = [&]() { /* all the push events here */ };
```
Then you just call it as `Push()` both before and after the event. Removes code duplication.

Can also make it look like:
```C
auto PushAndSignal = [&](std::string &ev) -> bool {
    Events::PushEventData("...", ...);
    ...
    return EventsSignal(ev, thisPtr->m_idSelf);
```

And then your core function is:

```C
if (PushAndSignal("NWNX_ON_USE_SKILL_BEFORE"))
{
    retVal = ...;
}

PushAndSignal("NWNX_ON_USE_SKILL_AFTER");
return retVal;
```

And same logic can be applied to anything that needs to push more than 1-2 pieces of event data.

Reply:
Added, that does make it look a lot more readable